common/searchattribute/name_type_map.go

```diff
@@ -0,0 +1,95 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package searchattribute
+
+import (
+	"fmt"
+
+	enumspb "go.temporal.io/api/enums/v1"
+	persistencespb "go.temporal.io/server/api/persistence/v1"
+)
+
+type (
+	NameTypeMap struct {
+		customSearchAttributes map[string]enumspb.IndexedValueType
+	}
+)
+
+func BuildIndexNameTypeMap(indexSearchAttributes map[string]*persistencespb.IndexSearchAttributes) map[string]NameTypeMap {
+	indexNameTypeMap := make(map[string]NameTypeMap, len(indexSearchAttributes))
+	for indexName, customSearchAttributes := range indexSearchAttributes {
+		indexNameTypeMap[indexName] = NameTypeMap{
+			customSearchAttributes: customSearchAttributes.GetCustomSearchAttributes(),
+		}
+	}
+	return indexNameTypeMap
+}
+
+func (m NameTypeMap) System() map[string]enumspb.IndexedValueType {
```

Comment:
`func (m NameTypeMap)` -> `func (m *NameTypeMap)`

Reply:
I wanted to pass `NameTypeMap` by value everywhere because it is just one pointer to the `map`. But few invocation requires `nil` because they might work w/o type map.

Reply:
NameTypeMap is immutable?

Reply:
Yes, this `map` pointer is never get changed. Even when content of the map is changed, it goes through database read/write, so it is new struct every time.